finetune_full.py

```diff
@@ -0,0 +1,212 @@
+from lit_llama.model import Block, LLaMA, LLaMAConfig
+from lit_llama.utils import save_model_checkpoint
+from lit_llama.tokenizer import Tokenizer
+from scripts.prepare_alpaca import generate_prompt
+
+
+eval_interval = 1000
+save_interval = 1000
+eval_iters = 100
+log_interval = 100
+devices = 4
+
+# Hyperparameters
+learning_rate = 3e-4
+batch_size = 128 / devices
+micro_batch_size = 4
+gradient_accumulation_steps = batch_size // micro_batch_size
+epoch_size = 50000  # train dataset size
+num_epochs = 5
+max_iters = num_epochs * epoch_size // micro_batch_size // devices
+weight_decay = 0.0
+block_size = 512
+warmup_steps = 100
+
+
+def main(
+    data_dir: str = "data/alpaca",
+    pretrained_path: str = "checkpoints/lit-llama/7B/lit-llama.pth",
+    out_dir: str = "out/regular/alpaca",
+):
+
+    auto_wrap_policy = partial(transformer_auto_wrap_policy, transformer_layer_cls={Block})
+    strategy = FSDPStrategy(auto_wrap_policy=auto_wrap_policy, activation_checkpointing=Block)
+
+    fabric = L.Fabric(accelerator="cuda", devices=4, precision="bf16-mixed", strategy=strategy)
+    fabric.launch()
+    fabric.seed_everything(1337 + fabric.global_rank)
+
+    if fabric.global_rank == 0:
+        os.makedirs(out_dir, exist_ok=True)
+
+    train_data, val_data = load_datasets(data_dir=data_dir)
+
+    config = LLaMAConfig.from_name("7B")
+    config.block_size = block_size
+
+    checkpoint = torch.load(pretrained_path)
+
+    with fabric.device:
+        torch.set_default_tensor_type(torch.HalfTensor)
+        model = LLaMA(config).bfloat16()
+        torch.set_default_tensor_type(torch.FloatTensor)
+        model.load_state_dict(checkpoint, strict=False) 
+
+    model = fabric.setup_module(model)
+
+    optimizer = torch.optim.AdamW(model.parameters(), lr=learning_rate)
+    optimizer = fabric.setup_optimizers(optimizer)
+
+    train(fabric, model, optimizer, train_data, val_data, out_dir)
+
+    # Save the final checkpoint at the end of training
+    save_model_checkpoint(fabric, model, os.path.join(out_dir, "lit-llama-regular-finetuned.pth"))
+
+
+def train(
+    fabric: L.Fabric,
+    model: torch.nn.Module,
+    optimizer: torch.optim.Optimizer,
+    train_data: np.ndarray,
+    val_data: np.ndarray,
+    out_dir: str,
+) -> None:
+    """The training loop.
+
+    Loosely based on the nanoGPT implementation: https://github.com/karpathy/nanoGPT.
+    """
+    step_count = 0
+    model.train()
+    print(model)
+
+    for iter_num in range(max_iters):
+
+        if step_count <= warmup_steps:
+            # linear warmup
+            lr = learning_rate * step_count / warmup_steps
+            for param_group in optimizer.param_groups:
+                param_group['lr'] = lr
+
+        t0 = time.time()
+
+        input_ids, targets = get_batch(fabric, train_data)
+        logits = model(input_ids)
+        loss = loss_fn(logits, targets)
+        fabric.backward(loss)
+
+        if (iter_num + 1) % gradient_accumulation_steps == 0:
```

Comment:
We can make the gradient accumulation a bit more efficient in the multi-gpu case here.
https://lightning.ai/docs/fabric/stable/advanced/gradient_accumulation.html
Could also be done in a follow up

Reply:
Didn't know about that! Cool! Added it (hopefully correctly)